### DISCUSSION TITLE
Syntax for including front matter data in plain markdown?

### DISCUSSION THREAD

**Discussion by ilg-ul** [facebook/docusaurus#8759] [Q&A]
I'm trying to migrate a web site from Jekyll, and my markdown pages have various variables defined in the front matter.

What is the correct syntax to access these variables in a plain `.md` file, since the Liquid `{{ variable }}` seems not accepted, nor the simple `{ variable }`.

To make things simple, let's assume a simple Jekyll blog page:

```md
---
title: My title
custom: something
---

{{ page.title }}
{{ page.custom }}
```

How would it look in Docusaurus MD?

Can we possibly edit <https://docusaurus.io/docs/markdown-features#front-matter> to show this?

Thank you!

**Comment by pranabdas** [comment]
You can try:

```
---
title: My title
custom: something
---

<h1>{ frontMatter.title }</h1>
<p>{ frontMatter.custom }</p>
<div>{ frontMatter.custom }</div>
```

**Reply by ilg-ul** [reply]
Hi @pranabdas, thank you for your suggestion.

Ok, that was too easy. Here is a real use case:

```
---
title:  xPack Ninja Build v1.11.1-2 released
summary: "Version **1.11.1-2** is a maintenance release, built with the latest build scripts."

ninja_version: 1.11.1
ninja_date: 30 Aug 2022
version: "1.11.1-2"
npm_subversion: 1
download_url: https://github.com/xpack-dev-tools/ninja-build-xpack/releases/tag/v1.11.1-2/
---

{{ page.summary }}
...
## Download

The binary files are available from GitHub [Releases]({ download_url }).
...
To install this specific version, use:

    ```
    xpm install @xpack-dev-tools/ninja-build@{{ page.version }}.{{ page.npm_subversion }}
    ```
...
The current version is based on:

- Ninja Build version {{ page.ninja_version }} from {{ page.ninja_date }}.
...
```

In other words, the substitution is part of some other elements, like code, lists, etc.

Isn't there any solution without having to add explicit html elements?

**Reply by ilg-ul** [reply]
I tried `<p>{ frontMatter.summary }</p>` in my page, and it was rendered, but as raw html, not as markdown (without the bold version). This is probably expected, which makes me think that this is not the desired solution.

Is there any way to do the substitution **before** the conversion from markdown to html?

**Reply by pranabdas** [reply]
Could you use html component instead of markdown?

```
<ul>
  <li>some text { frontMatter.custom }</li>
</ul>

<pre>
  some command { frontMatter.custom }
</pre>
```

**Reply by ilg-ul** [reply]
> Could you use html component instead of markdown?

Yeah, but that would contradict the main idea of using markdown as an easier way to enter content.

**Reply by pranabdas** [reply]
I do not know if it is possible with MDX. The markdown is parsed by <https://github.com/mdx-js/mdx>. You may look into their documentation. Note that Docusaurus currently uses version 1 of MDX and planning to upgrade to v2.

**Reply by ilg-ul** [reply]
I checked the MDX documentation before, but things are very complicated and I could not identify a solution. :-(

The problem seems to be the different meaning of _markdown_ in MDX, which is closer to HTML than in usual (GitHub) markdown.
 
The ideal solution would be to process the substitutions via LiquidJS before any other processing, but I don't know if this can be done, even with a MDX plug-in...

**Reply by ilg-ul** [reply]
> Docusaurus currently uses version 1 of MDX and planning to upgrade to v2.

I checked the MDX documentation again, and it seems that substitutions are now possible in markdown elements:

- https://mdxjs.com/docs/using-mdx/#props

@pranabdas, do you know the schedule for the upgrade to MDX v2? The sites I plan to migrate to Docusaurus have quite a lot of such substitutions (common to Jekyll), and perhaps it would be easier to wait.

**Reply by pranabdas** [reply]
I do not know, you may follow the developments here https://github.com/facebook/docusaurus/pull/8288

**Reply by ilg-ul** [reply]
It is work in progress, but it looks like it might take a while.

**Reply by slorber** [reply]
MDX 2 is the top-priority PR for me but it is not an easy one 🤪 still working on it but it's coming close to the end.

> The problem seems to be the different meaning of markdown in MDX, which is closer to HTML than in usual (GitHub) markdown.

GitHub does not have this kind of `{{ var }}` syntax, this is proprietary from Jekyll and not in commonmark

Note technically you can create a Remark plugin to invent your own variable syntax

Some do that here if you want some inspiration: https://github.com/facebook/docusaurus/issues/395

**Reply by ilg-ul** [reply]
> GitHub does not have this kind of {{ var }} syntax, this is proprietary from Jekyll and not in commonmark

Right. 

In fact my problem is not the different syntax, it's the different functionality, more precisely the moment when the substitutions are performed.

Jekyll first runs the file through the template engine, which performs all substitutions, and then parses the markdown syntax and generates the HTML code.

With Docusaurus and MDX 1, it looks like variables are only accepted in the React HTML-like syntax, and not in the markdown text.

In the MDX 2 documentation I encountered an example where variables were used in a H1 element:

- https://github.com/mdx-js/specification#inline

@slorber, can you confirm that with MDX 2 variables can also be used in markdown? (with that slightly more complex syntax)

The point is that most of my pages (hundreds of them) have such variables, and rewriting all references in React HTML would be tedious. If MDX 2 simplifies this process, I'd probably wait for it before starting the migration.

**Reply by slorber** [reply]
@ilg-ul MDX compiles Markdown to React components, unlike Jekyll

`# Hello, from <>{props.from}</>!`

Becomes:

```tsx
function _createMdxContent(props) {
  const _components = Object.assign({
    h1: "h1"
  }, props.components);
  return _jsxs(_components.h1, {
    children: ["Hello, from ", _jsx(_Fragment, {
      children: props.from
    }), "!"]
  });
}
```

Try it with the mdx playground: https://mdxjs.com/playground/

There is no actual substitution in MDX, the code variable remain code variables. 

The TOC is accessible as variables this way (we automatically generate the frontMatter export for you)

```md
export const frontMatter = {name: "Sebastien"}

Hello {frontMatter.name}
```

This is compiled to: 

```jsx
const frontMatter = {
  name: "Sebastien"
};

function _createMdxContent(props) {
  const _components = Object.assign({
    p: "p"
  }, props.components);
  return _jsxs(_components.p, {
    children: ["Hello ", frontMatter.name]
  });
}
```

If you want a real substitution, this means you want this code after mdx compilation:

```tsx
function _createMdxContent(props) {
  const _components = Object.assign({
    p: "p"
  }, props.components);
  return _jsx(_components.p, {
    children: "Hello Sebastien"
  });
}
```

This is not part of MDX. You will have to implement it yourself with a custom or third-party remark plugin, cf the link I provided above for examples of how it can be done.

---

When in doubt with MDX, try to run code in the playground and see how what you try to do works.

- MDX 1 playground: https://mdx-git-renovate-babel-monorepo-mdx.vercel.app/playground/
- MDX 2 playground: https://mdxjs.com/playground/

`# Hello, from <>{props.from}</>!` does not work in MDX 1

Although it will work in MDX 2, I don't recommend it for this specific case (heading), because it prevents Docusaurus from computing an appropriate label for the table-of-contents. In this case, I'd recommend using a hardcoded label for now, or a remark plugin with real build-time substitution.

**Reply by ilg-ul** [reply]
Thank you, Sébastien @slorber.

The link to the MDX 2 playground was useful; I tried variables in a markdown code part, and they were not processed.

```
	```bash
	xpm install @xpack-dev-tools/ninja-build@<>{frontMatter.version}</>.<>{frontMatter.npm_subversion}</>
	```
```

Same for list elements:

```
	- Docus Mock version <>{ frontMatter.ninja_version }</> from <>{ frontMatter.ninja_date }</>.
```

I don't understand yet if a remark plugin can help in this case, but the sites I plan to migrate are full of such parametrised markdown and manually changing them into HTML is not exactly fun.

**Reply by slorber** [reply]
MDX 2 works for variables in list elements:

<img width="992" alt="CleanShot 2023-03-16 at 12 09 08@2x" src="https://user-images.githubusercontent.com/749374/225599330-73e09361-5259-4ddf-a8ab-01498c3f229b.png">

MDX doesn't do substitution and leaves code blocks as pure strings. If you want variables in code blocks you'd need to use Remark plugins for a real substitution system. You can also try to swizzle the code block and transform that string yourself at runtime before passing it to the syntax highlighter.

There's not only 1 single solution to solve this problem.

**Reply by ilg-ul** [reply]
> MDX 2 works for variables in list elements

Great! 

I'll wait for the next Docusaurus release and resume the migration. 

For the code parts I'll probably have to invent some sed scripts to convert them from markdown to html.

**Comment by ilg-ul** [comment]
Punctual question:

Since direct conversion seems not possible, what would be the Docusaurus/React recommended solution to implement something like this:

```
    ```sh
    xpm install @xpack-dev-tools/ninja-build@{{ page.version }}.{{ page.npm_subversion }}
    ```
```

In other words, a piece of parametrised shell code.

**Comment by pranabdas** [comment]
A solution that works:

```html
---
version: 1
npm_subversion: 2
---

<pre className="language-bash">
xpm install @xpack-dev-tools/ninja-build@{frontMatter.version}.{frontMatter.npm_subversion}
</pre>
```

**Reply by ilg-ul** [reply]
I confirm, this worked.

**Comment by ilg-ul** [comment]
I found a short reference in the MDX pages:

- https://github.com/mdx-js/specification#inline

It looks like empty elements can be used to brace substitutions in regular text. It doesn't work in lists or code elements, which must be re-written completely in HTML.
